### DISCUSSION TITLE
Is it possible to 'apply' a PrimeFaces datepicker to an input field from Javascript.

### DISCUSSION THREAD

**Discussion by tbee** [primefaces/community#2823] [PrimeFaces] [answered]
After some Googling and documentation reading I ended up here:

[DatePicker Javascript API docs](https://primefaces.github.io/primefaces/jsdocs/classes/src_PrimeFaces.PrimeFaces.widget.DatePicker-1.html)

But my Javascript skills are not impressive enough to figure out how to get that to work.

AI tells me this should work

```
PrimeFaces.widget.DatePicker.init({
  id: 'myDatePicker',
  widgetVar: 'myDatePickerWidget',
  showIcon: true,
  dateFormat: 'mm/dd/yy'
});
```

But it isn't and the code seems not correct. Init is not a factory method?

**Comment by melloware** [comment]
@tbee whatever you are trying to do I don't recommend it.   Can you explain your use case and why not just use a `<p:datePicker>`?

**Comment by tbee** [comment]
Well, we're working on a 20+ year old webapp that has all kinds of web technology mixed together. In this case it is a JSF/PF page combined with a load of jquery. The current code uses a jquery-ui datepicker call to upgrade an input to a datepicker. Since we're trying to standarize on PF's datepicker, and since that has a Javascript API, I was hoping to replace that jquery-ui call with a PF one instead of having to rewrite the whole page...

**Reply by melloware** [reply]
Ahhh that won't be easy.  It may be possible but I would not recommend it. The proper way to create a PF widget is this way...

```js
pf.cw("DatePicker","widget_form_basic",{id:"form:basic",inline:false,userLocale:"en_US",dateFormat:"m\/d\/yy",selectionMode:"single",showOnFocus:true,appendTo:"@(body)",timeInput:false});``

pf.cw stands for Create Widget

**Reply by tbee** [reply]
Thanks! I'll give it a try Thursday. If it works, it works. And if it breaks sometime in the future, we can still rewrite the page then.

Ahm, that 'pf', is that a variable, andif so, how is it initialized?

**Reply by melloware** [reply]
Its from PrimeFaces and its in the PrimeFace sglobal namespace `var pf=window.PrimeFaces;`

Just go here: https://www.primefaces.org/showcase/ui/input/datepicker/datePicker.xhtml and do View Page source all the JS code that loads is at the bottom of the HTML.

**Reply by tbee** [reply]
Ah, yes, that I know. Thanks!

**Comment by tbee** [comment]
At the first attempt this fails because `this.jq.data()` has no values, it should contain primeDatePicker and primefacesOverlayTarget. Something is going wrong in the construction, trying to find where those two values are set.

```
Uncaught TypeError: Cannot read properties of undefined (reading 'primeDatePicker')
    at d.init (datepicker.js.faces?ln=primefaces&v=14.0.4&v2024.8.0:129:180)
    at d.init (core.js.faces?ln=primefaces&v=14.0.4&v2024.8.0:117:383)
    at new d (core.js.faces?ln=primefaces&v=14.0.4&v2024.8.0:117:157)
    at Object.createWidget (core.js.faces?ln=primefaces&v=14.0.4&v2024.8.0:27:298)
    at Object.cw (core.js.faces?ln=primefaces&v=14.0.4&v2024.8.0:27:25)
    at addDatepicker (careOrderRequestContent.js:213:8)
    at HTMLDocument.<anonymous> (modify:221:37)
    at e (jquery.js.faces?ln=primefaces&v=14.0.4&v2024.8.0:2:27028)
    at t (jquery.js.faces?ln=primefaces&v=14.0.4&v2024.8.0:2:27330)
```

**Reply by melloware** [reply]
Make sure you have the DatePicker JS code loaded on the page?

**Reply by tbee** [reply]
Yes. But I found it already (as usual immediately after creating the comment); it turns out to be a naming thing in the surrounding span.

**Reply by tbee** [reply]
Well, how about that? So far so good, I've got a popup. 😄

**Reply by melloware** [reply]
Amazing can you post your final solution here when you get it all working?

**Reply by tbee** [reply, marked as answer]
It is working. 😀 

So key takeaways: 

- Duplicate the HTML as generated by PrimeFaces: both the span and the input, make sure to keep the class attributes, but e.g. the aria attributes can be stripped.
- The naming of the span and input are important: they both need the "form:" prefix, and (of course) the input needs the "_input" suffix.
- And then it is just a simple call, where the elementName is the name including the "form:" prefix, but without the "_input" suffix.

```
    window.PrimeFaces.cw("DatePicker", elementName + "Widget", {
        id:elementName,
        inline:false,
        userLocale:"nl",
        dateFormat:"dd-mm-yy",
        showIcon:true,
        selectionMode:"single",
        showOnFocus:false,
        appendTo:"@(body)",
        showButtonBar:true,
        timeInput:false
    });
```

Thanks for helping! We're gonna let this sit and see how much trouble it causes, until it warrants a rewrite (or is replaced anyhow).
